### ISSUE TITLE
Metadata needs to contain arch-specific differences 

### ISSUE THREAD

**Issue by sotteson1** [microsoft/win32metadata#384]
We either need arch-specific types (e.g. x86 and x64) in the same .winmd, or we need multiple .winmd's for multiple architectures. Eventually we may want ARM and ARM64 support too. For now the metadata builds for 64-bit only.

**Comment by kennykerr** [comment]
How about a common arch-neutral winmd supplemented with arch-specific winmds? Something like this:
	
- Windows.Win32.winmd (almost everything lives here)
- Windows.Win32.x86.winmd
- Windows.Win32.x64.winmd
- Windows.Win32.Arm.winmd
- Windows.Win32.Arm64.winmd

**Comment by timsneath** [comment]
This approach makes me nervous for portable languages and frameworks like Dart/Flutter. The idea that I have to parse through five separate scopes to figure out the union of metadata is close to a dealbreaker for me. 

If I were entitled to a vote, I'd prefer for an attribute-based approach if practical, rather than generating so many different files. 

Imagine the matrix of generated files if you wind up also generating a file for each Windows 10 release that introduces new APIs... Do we wind up with Windows.Win32.21337.arm64.winmd?

**Comment by kennykerr** [comment]
Having multiple winmd files has always been a normal way of dealing with metadata. The [WinRT metadata package](https://www.nuget.org/packages/Microsoft.Windows.SDK.Contracts) for the Windows SDK ships with over 80 individual winmd files. It doesn't really matter how many files there are. A language projection must always be able to aggregate up all the metadata and cross reference as needed.

What's new here is having arch-specific winmd files. Up to now we've only ever had arch-neutral winmd files. I'd also prefer an attribute-based approach, but Steve tells me that may be a lot more work.

**Comment by timsneath** [comment]
Yeah, the difference is that if I load `Windows.Globalization.winmd`, I know that `Windows.Globalization.Calendar` is fully defined there. I don't have to also check some unknown number of sister files to see whether that definition is incomplete or even inaccurate. 

The Windows API even assumes that it's possible to resolve a Windows Runtime type to a metadata file, for example `RoGetMetadataFile`:
https://docs.microsoft.com/en-us/windows/win32/api/rometadataresolution/nf-rometadataresolution-rogetmetadatafile

An attribute approach feels like a more generally useful scenario, since you can both express something like:

```
[MinimumWindowsVersion("10.0.19041.5")]
```

as well as

```
[ArchitecturesSupported(x64, arm64)]
```

**Comment by kennykerr** [comment]
I agree about the attribute approach and it is consistent with many other aspects of the Win32 metadata that relies on attributes. 

This is not really relevant to this issue, but the following statement is not accurate:

> Yeah, the difference is that if I load Windows.Globalization.winmd, I know that Windows.Globalization.Calendar is fully defined there. 

If by "fully defined" you mean the TypeDef is present then yes that's probably true. But if by "fully defined" you mean everything you need to know about `Calendar` to define all aspects of the type, then no that's not necessarily true. For example, the `Calendar` class has a few constructors with collection parameters and those collection types are defined in the `Windows.Foundation.Collections` namespace that in all likelyhood are defined in a different winmd.

**Comment by sotteson1** [comment]
> What's new here is having arch-specific winmd files. Up to now we've only ever had arch-neutral winmd files. I'd also prefer an attribute-based approach, but Steve tells me that may be a lot more work.

The work isn't in using attributes vs putting them into different winmd files. Either way the work is in diffing and separating classes/structs/enums/methods to figure out which are OK for both, which are x86 only and which are x64 only. If I could build two complete winmd files, once for x86 and once for x64, then it's trivial and I could do that in a day, but Kenny says he can't afford to have another 12 MB winmd file.

**Comment by timsneath** [comment]
@kennykerr Right -- although the `Calendar` class is defined there, its dependencies may not be. 

But the difference is that I _know_ with confidence that I can't fully resolve a method based on the information I have, and I can just go call `RoGetMetadataFile` again to find the collection type. Versus the above approach where I might _or might not_ have sufficient information to completely define the projection.

And @sotteson1 yes -- that might be a showstopper for me also, because of maximum package upload size limits.

**Comment by kennykerr** [comment]
Yes, going from 12MB to 36MB (x86,x64,Arm64) just because < 1% of APIs have arch-specific definitions is very problematic for environments where the winmd files must be packaged up as part of the project. 

C++ and .NET project dependencies are enormous. Those developers are used to that kind of bloat with years of multi-gigabyte installs of VS and the SDK. Rust dependencies are measured in KBs. Here is a popular "large" crate weighing in at 232 KB. https://crates.io/crates/syn Most are much smaller. At first the windows crate seems pretty lean at only 18.7 KB https://crates.io/crates/windows but then you drill into its dependencies and see it pulls in an extra 5.7 MB (compressed) for windows_gen https://crates.io/crates/windows_gen/0.6.0 Oh, and Rust must support offline builds - so we can't just have it download on demand. It has to be packaged up to build in an offline manner.

As it is, one of the main complaints is that it's too big and slow to download and compile. Imagine trying to convince a developer to take on 36MB of binary dependencies just to call MessageBoxA. If this is something that truly only matters to Rust, I can try and write a post processing tool to do it after you publish the arch winmds. But I suspect it will also be an issue for JIT languages like JavaScript and @timsneath mentioned it may also be an issue for Dart.

**Comment by timsneath** [comment]
(For Dart, it's not a runtime issue for a couple of reasons -- both because we perform tree shaking during compilation on any unused typedefs, and because the winmd is marked as a development-only dependency. But yes, I empathize with Kenny!)

**Comment by mikebattista** [comment]
It sounds like, though, the ideal experience for projection authors and consumers is one winmd with attributes to handle the <1% arch-specific cases. Is that right?

**Comment by mikebattista** [comment]
And would it make sense to only add the attribute on the <1% cases? So omission of the attribute means works for everything, and as a projection author, if you encounter the attribute, you can change your projection per-arch accordingly?

**Comment by kennykerr** [comment]
Yes.

**Comment by mikebattista** [comment]
Ok. That feels achievable.

**Comment by AraHaan** [comment]
I would prefer them to opt into arch specific definitions to stuff only if they want it.

**Comment by mikebattista** [comment]
Could you clarify what you mean? I didn't quite parse what you were saying there.

**Comment by AraHaan** [comment]
have them opt into arch specific generation by them using ``[assembly: ArchitecturesSupported()]`` where they list the arch's they want the source generator to generate for.

**Comment by AArnott** [comment]
> Or if they do not provide it then assume generic arch like currently.

We don't assume generic arch. The metadata is only correct today for 64-bit architectures. Most APIs are in fact equivalent between the two, but those that aren't, are wrong for 32-bit. The metadata needs to represent enough for all architectures to be generated correctly. 

The method by which a source generator selects which architecture to generate code for will be a projection-level decision rather than one discussed in this repo.

**Comment by MikeYeager** [comment]
Not sure if this extra information helps. I just updated to 0.1.422-beta and not only does the OPENFILENAMEW still not work due to the wrong size being returned, it started giving me an error using sizeof() on the type. I've tried to calculate the actual size, but can't drill into the generated type far enough to see what it's actually made up of. sizeof() used to return 140 and I think we decided it should have been returning 152. Neither size works now.
![image](https://user-images.githubusercontent.com/7063813/114452108-4e3d3a80-9b95-11eb-94a9-393b981c42e2.png)
